helion/language/loops.py

```diff
@@ -449,6 +451,177 @@ def _(state: CodegenState) -> ast.AST:
     return _codegen_loop_helper(state)
 
 
+def _to_int(value: int | torch.Tensor | None) -> int | None:
+    """Convert tensor values to int."""
+    if value is None:
+        return None
+    if isinstance(value, torch.Tensor):
+        return int(value.item())
+    return int(value)
+
+
+def _normalize_to_list(
+    value: int | torch.Tensor | list[int | torch.Tensor],
+) -> list[int | torch.Tensor]:
+    """Convert single values to lists for uniform handling."""
+    if isinstance(value, (list, tuple)):
+        return list(value)
+    return [value]
+
+
+def _normalize_begin_end_ref(
+    begin_or_end: int | torch.Tensor | list[int | torch.Tensor],
+    end_or_none: int | torch.Tensor | list[int | torch.Tensor] | None = None,
+) -> tuple[
+    int | torch.Tensor | list[int | torch.Tensor],
+    int | torch.Tensor | list[int | torch.Tensor],
+]:
+    if end_or_none is not None:
+        # Two positional args: begin_or_end is begin, end_or_none is end
+        return begin_or_end, end_or_none
+    # One positional arg: begin_or_end is end, begin defaults to 0
+    end = begin_or_end
+    if isinstance(end, (list, tuple)):
+        begin = cast("int | torch.Tensor | list[int | torch.Tensor]", [0] * len(end))
+    else:
+        begin = 0
+    return begin, end
+
+
+def _get_block_size_from_config(num_dims: int) -> list[int]:
+    """Get block size from ref mode config."""
+    env = CompileEnvironment.current()
+    block_sizes_sequence = env.config_spec.block_sizes
+
+    # Get the default configuration values
+    if len(block_sizes_sequence) == 0:
+        return []
+
+    # Get default values from each BlockSizeSpec
+    block_sizes = []
+    for spec in block_sizes_sequence:
+        # Get the default value from the fragment
+        fragment = spec._fragment(env.config_spec)
+        default_value = fragment.default()  # Call the method
+        block_sizes.append(default_value)
+
+    if len(block_sizes) < num_dims:
+        from .. import exc
+
+        raise exc.InvalidConfig(
+            f"Not enough values for config['block_sizes']: expected {num_dims} block sizes "
+            f"(one for each tiled dimension), got {len(block_sizes)}. "
+            f"Did you forget to specify block sizes for all your hl.tile() dimensions?"
+        )
+
+    # Return exactly the needed number of block sizes
+    return block_sizes[:num_dims]
+
+
+def _determine_block_sizes(
```

Comment:
Why aren't we always generating a single block?  I was thinking that `hl.tile()` would always return just a single tile covering the entire iteration space.

Processing tiles eagerly will be very slow, since you will be launching one kernel per tile and each of those kernels will only have enough work for a single SM so it will have <5% GPU utilization.

Maybe the question here is this for debugging, or is this for estimating eager performance.

Once we have a `torch.compile` version, the single-block one will be the only one that makes sense.